### DISCUSSION TITLE
2주차 사전 과제, 김우민

### DISCUSSION THREAD

**Discussion by gleaming9** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#26] [General]
### 1. 일상생활에서 “상태(state)” 라는 개념이 사용되는 예시를 1가지 작성해주세요. 각 예시에서 “상태” 가 어떻게 변하고, 그 변화가 어떤 결과를 가져오는지 적어주시면 됩니다.

신호등을 예시로 들 수 있다.
1. 빨간불 : 정지
2. 노란불 : 주의
3. 초록불 : 진행

빨간불은 차량을 멈추게하고 노란불은 운전자들로 하여금 경각심을 부여하며, 초록불은 차량을 진행하게 만든다. 이러한 변화가 교통 흐름의 질서를 가져온다.

### 2. 자신이 주로 사용하는 웹 페이지의 이름과, 그 웹 페이지에서 사용자의 입력에 따라 즉시 반응하는 요소들의 예를 3가지 작성해주세요. 각 요소가 어떤 사용자 행동에 반응하고, 어떻게 변화하는지 설명해주세요.
 주로 사용하는 웹 페이지로는 구글이 있다.
1. 검색 자동 완성 : 사용자가 검색어를 입력할 때, 텍스트와 관련된 검색어를 분석해 관련 검색어를 추천하고 검색어를 클릭하면 검색 결과 페이지를 열어준다.
2. 검색 결과 : 사용자가 검색어를 클릭하거나 엔터를 칠 때, 검색 결과를 보여준다.
3. 검색 도구 : 검색 페이지에서 도구 버튼을 클릭하면, 검색 결과를 세부적으로 조정할 수 있는 선택지를 보여준다.

### 3. JavaScript로 작성된 다음 코드를 실행했을 때 console.log(x.count) 의 결과를 작성하고, 그 결과가 나온 이유를 작성해주세요.
```
let x = { count: 0 };
let y = x;
y.count = 1;
console.log(x.count);
```
console.log(x.count)의 결과값은 1이 된다. 2번째 줄을 통해서 y는 x가 참조하고 있는 객체를 동일하게 참조하게 된다. 따라서 y를 통해서 데이터를 변환하면 같은 객체를 참조하고 있는 x를 통해서 값을 확인했을 때, 동일하게 달라진 결과값을 확인할 수 있다.

### 4. 2주차 스터디 도중 props의 특성 중 “불변성(immutability)”에 대해서 설명드렸습니다. 프로그래밍에서 불변성을 지키는 것이 중요할까요? 자신의 의견과 이유를 적어주세요.
 불변성을 지키는 것은 코드의 예측 가능성을 높이고, 디버깅을 쉽게 하며, 동시성 문제를 줄여주기 때문에 불변성을 지키는 것을 권장해야 한다고 생각한다.

### 5. JavaScript로 작성된 다음 코드를 실행했을 때 어떤 순서로 콘솔에 출력될 지 예측해보세요. 자신이 예측했던 결과와 그 이유, 그리고 실제 출력된 결과를 작성해주세요.

```
console.log('시작');

setTimeout(() => {
  console.log('3초 후');
}, 3000);

setTimeout(() => {
  console.log('1초 후');
}, 1000);

console.log('끝');
```
'시작'이 출력된 후 3초가 지나고 '3초 후'가 출력되고, 1초가 지나고 '1초 후'가 출력되고 바로 '끝'이 출력될 것이라고 예측했다. 자바스크립트가 비동기 함수를 처리하는 방식을 간과했기 때문이다. 실제로는 '시작', '끝', '1초 후', '3초 후' 순서로 출력되었다. 자바스크립트는 이벤트 루프를 통해 비동기 작업을 처리하여 블로킹 없이 코드가 실행된다.

**Comment by MrMirror21** [comment]
작성해주신 사전과제 잘 봤습니다. 전체적으로 답변 훌륭하게 잘 답변해주셨네요. 풀이형 문제에서 다른 답을 생각하셨더라도 정답의 원리에 관해 찾아보신 점이 인상적입니다. 사전과제 작성이 지난 3주차 스터디 내용을 이해하는데 도움이 되셨으면 좋겠습니다. 열심히 참여해주셔서 감사합니다 🙌
